build.gradle.kts

```diff
@@ -0,0 +1,277 @@
+        val moduleName = "org.jetbrains.annotations" // this module's name
+        val compileKotlinJvm by getting(KotlinCompile::class)
+        val compileJava by getting(JavaCompile::class)
+        val sourceDir = file("src/jvmMain/moduleInfo/")
+        val targetDir = compileKotlinJvm.destinationDirectory.map { it.dir("../moduleInfo/") }
+
+        // Use a Java 11 compiler for the module info.
+        javaCompiler.set(project.javaToolchains.compilerFor { languageVersion.set(JavaLanguageVersion.of(11)) })
+
+        // Always compile kotlin and java classes before the module descriptor.
+        dependsOn(compileKotlinJvm)
+        dependsOn(compileJava)
+
+        // Add the module-info source file.
+        source(sourceDir)
+
+        // Set the task outputs and destination dir
+        outputs.dir(targetDir)
+        destinationDirectory.set(targetDir)
+
+        // Configure JVM compatibility
+        sourceCompatibility = JavaVersion.VERSION_1_9.toString()
+        targetCompatibility = JavaVersion.VERSION_1_9.toString()
+
+        // Set the Java release version.
+        options.release.set(9)
+
+        // Ignore warnings about using 'requires transitive' on automatic modules.
+        // not needed when compiling with recent JDKs, e.g. 17
+        options.compilerArgs.add("-Xlint:-requires-transitive-automatic")
+
+        // Patch the compileKotlinJvm output classes into the compilation so exporting packages works correctly.
+        options.compilerArgs.addAll(listOf("--patch-module", "$moduleName=${compileJava.destinationDirectory.get()}"))
+
+        // Use the classpath of the compileKotlinJvm task.
+        // Also ensure that the module path is used instead of classpath.
+        classpath = compileKotlinJvm.libraries
+        modularity.inferModulePath.set(true)
+    }
+
+    // Configure the JAR task so that it will include the compiled module-info class file.
+    val jvmJar by existing(Jar::class) {
+        manifest {
+            attributes("Multi-Release" to true)
+        }
+        from(compileJavaModuleInfo.map { it.destinationDirectory }) {
+            into("META-INF/versions/9/")
+        }
+    }
+
+    val allMetadataJar by existing(Jar::class) {
+        archiveClassifier.set("all")
+    }
+
+    val javadocJar by creating(Jar::class) {
+        from(javadoc)
+        archiveClassifier.set("javadoc")
+    }
+}
+
+nexusPublishing {
+    repositories {
+        sonatype {
+            username.set(publishingUser)
+            password.set(publishingPassword)
+        }
+    }
+}
+
+publishing {
+    val artifactBaseName = base.archivesName.get()
+    configureMultiModuleMavenPublishing {
+        val rootModule = module("rootModule") {
+            mavenPublication {
+                artifactId = artifactBaseName
+                groupId = "org.jetbrains"
+                configureKotlinPomAttributes(packaging = "jar")
+                artifact(tasks.getByName("javadocJar"))
```

Comment:
Better to configure an outgoing configuration with proper attributes for javadoc, include that artifact in it, and publish it as a variant in the root module.
See the example of such variant here: https://repo1.maven.org/maven2/org/jetbrains/kotlin/kotlin-gradle-plugin/2.0.20/kotlin-gradle-plugin-2.0.20.module